src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -18955,5 +19045,160 @@ public void ExplicitCast_StaticType()
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "default(T)").WithLocation(5, 22),
+                // (8,15): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         T t = (T)NullableObject(); // warn: T may be non-null
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(T)NullableObject()").WithLocation(8, 15),
+                // (14,22): warning CS8625: Cannot convert null literal to non-nullable reference.
+                //     static U F1() => default; // warn: T may be non-null
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "default").WithLocation(14, 22),
+                // (15,22): warning CS8625: Cannot convert null literal to non-nullable reference.
+                //     static U F2() => default(U); // warn: U may be non-null
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "default(U)").WithLocation(15, 22),
+                // (18,15): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         U u = (U)NullableObject(); // warn: U may be non-null
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(U)NullableObject()").WithLocation(18, 15),
+                // (18,15): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         U u = (U)NullableObject(); // warn: U may be non-null
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "(U)NullableObject()").WithLocation(18, 15));
+        }
+
+        [Fact]
+        public void UnconstrainedTypeParameter_MayBeNullable()
+        {
+            var source =
+@"class C
+{
+    static void F(object o)
+    {
+    }
+    static void F1<T>(T t1)
+    {
+        F(t1);
+        t1.ToString();
+    }
+    static void F2<T>(T t2) where T : struct
+    {
+        F(t2);
+        t2.ToString();
+    }
+    static void F3<T>(T t3) where T : class
+    {
+        F(t3);
+        t3.ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,11): warning CS8604: Possible null reference argument for parameter 'o' in 'void C.F(object o)'.
+                //         F(t1);
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "t1").WithArguments("o", "void C.F(object o)").WithLocation(8, 11),
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         t1.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t1").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void UnconstrainedTypeParameter_Return()
+        {
+            var source =
+@"class C
+{
+    static object? F0<T>(T t) => t;
+    static object F1<T>(T t) => t;
+    static T F2<T>(T t) => t;
+    static T F3<T, U>(U u) where U : T => u;
```

Comment:
>static T F3<T, U>(U u) where U : T => u; [](start = 4, length = 40)

This scenario and below it might be also worth testing cases when return type is object and object?